### DISCUSSION TITLE
`BackgroundTask` runs the tasks before response is sent if the return statement is another function call

### DISCUSSION THREAD

**Discussion by BelloAngo** [fastapi/fastapi#11850] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
@router.post(...)
async def validator_create(
    validator_in: create_schemas.ValidatorCreate,
    current_admin: CurrentAdmin,
    db: DatabaseSession,
    background_tasks: BackgroundTasks,
):
    """This endpoint creates a new validator"""
    # Check if current admin is a super admin
    ...
    background_tasks.add_task(
        sendgrid.send_email,
        subject="Your iBloom Validator Login Credentials",
        to=validator.email,
        body=content,
    )

    background_tasks.add_task(
        services.generate_validator_worksheets, validator_id=validator.id, db=db
    )
    return {"data": await format_validator(validator=validator)}
```

### Description

When the above endpoint is called, it runs the background tasks before the response is set.
However when i declare a vaiable i.e. `response = {"data": await format_validator()}` it will work normally then return that, it will then work properly.

NOTE: i saw that the logs in the background tasks only appear after the logs show `/validator 200 OK`

### Operating System

Windows

### FastAPI Version

0.111.0

### Pydantic Version

2.7.1

### Python Version

3.12.4

**Comment by YuriiMotov** [comment]
Can you create a minimal code example that we can run and see the behavior you mentioned?
How did you understand that 'it runs the background tasks before the response is set'?

**Comment by BelloAngo** [comment]
Hello, i have been trying to replicate it but i am facing some troubles.

After trying some things i noticed that the request isnt run during the request but after, so the background tasks work but the frontend waits in a way till after the background task finishes running before cutting the connection. Here are some logs:

![image](https://github.com/user-attachments/assets/0d072ff9-7c4b-497e-9b6b-109d4a35a225)

I timed the time each call takes, so the frontend keeps the request as "ongoing" in a way i.e wont close connection or something until the generator finishes then it will close.

NOTE: the generator is an async function that calls a bunch of other async functions and it also has a couple of loops

**Reply by YuriiMotov** [reply]
I can't help you unless you show the code that shows incorrect behavior.

I just created a simple endpoint with background tasks and it works as intended - background tasks are started only after sending the response. Browser receives response in 0.501 sec:

```
import asyncio
from time import sleep
import fastapi

app = fastapi.FastAPI()

def bg_task_1():
    print("bg_task_1 started")
    sleep(10)
    print("bg_task_1 finished")

def bg_task_2():
    print("bg_task_2 started")
    sleep(10)
    print("bg_task_2 finished")

async def format_validator(validator: str):
    await asyncio.sleep(0.5)
    return validator


@app.get("/")
async def validator_create(
    background_tasks: fastapi.BackgroundTasks,
):
    background_tasks.add_task(
        bg_task_1
    )
    background_tasks.add_task(
        bg_task_2
    )
    return {"data": await format_validator(validator="some data")}
```

**Reply by BelloAngo** [reply]
Alright i will keep looking into it, thank you
